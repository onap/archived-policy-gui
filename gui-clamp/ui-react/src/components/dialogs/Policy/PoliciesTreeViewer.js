/*-
 * ============LICENSE_START=======================================================
 * ONAP POLICY-CLAMP
 * ================================================================================
 * Copyright (C) 2021 AT&T Intellectual Property. All rights reserved.
 * Modifications Copyright (C) 2022 Nordix Foundation.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

import React from 'react'
import TreeView from '@material-ui/lab/TreeView';
import TreeItem from '@material-ui/lab/TreeItem';
import FolderIcon from '@material-ui/icons/Folder';
import FolderOpenIcon from '@material-ui/icons/FolderOpen';
import DescriptionIcon from '@material-ui/icons/Description';


export default class PoliciesTreeViewer extends React.Component {

    separator = ".";

    nodesList = new Map();

    constructor(props, context) {
        super(props, context);
        this.createPoliciesTree = this.createPoliciesTree.bind(this);
        this.handleTreeItemClick = this.handleTreeItemClick.bind(this);
        this.buildNameWithParent = this.buildNameWithParent.bind(this);

    }

    state = {
        policiesTreeData: this.createPoliciesTree(this.props.policiesData),
    }

    componentDidUpdate(prevProps) {
        if (prevProps.policiesData !== this.props.policiesData) {
            this.setState({policiesTreeData: this.createPoliciesTree(this.props.policiesData)})
        }
    }

    createPoliciesTree(policiesArray) {
        console.log('createPoliciesTree called')
        // put my policies array in a Json
        let nodeId = 1;
        let root = {id: nodeId, policyCount: 0, name: "ROOT", children: [], parent: undefined};
        this.nodesList.set(nodeId++, root);

        if (policiesArray !== null && policiesArray.forEach !== undefined) {

            policiesArray.forEach(policy => {
                let currentTreeNode = root;
                policy[this.props.valueForTreeCreation].split(this.separator).forEach((policyNamePart, index, policyNamePartsArray) => {
                    let node = currentTreeNode["children"].find(element => element.name === policyNamePart);
                    if (typeof (node) === "undefined") {
                        node = {
                            id: nodeId,
                            policyCount: 0,
                            children: [],
                            name: policyNamePart,
                            parent: currentTreeNode
                        };
                        this.nodesList.set(nodeId++, node);
                        currentTreeNode["children"].push(node);
                    }
                    if ((index + 1) === policyNamePartsArray.length) {
                        ++currentTreeNode["policyCount"];
                    }
                    currentTreeNode = node;
                })
            });
        }
        return root;
    }

    buildNameWithParent(node) {
        let nameToBuild = node.name;
        if (node.parent !== undefined) {
            nameToBuild = this.buildNameWithParent(node.parent) + this.separator + node.name;
        }
        return nameToBuild;
    }

    handleTreeItemClick(event, value) {
        let fullName = this.buildNameWithParent(this.nodesList.get(value[0])).substring(5);
        this.props.policiesFilterFunction(fullName);
    }

    renderTreeItems(nodes) {
        return (<TreeItem key={nodes.id} nodeId={nodes.id} label={nodes.name + "(" + nodes.policyCount + ")"}
                          onNodeSelect={this.handleTreeItemClick}>
            {
                Array.isArray(nodes.children) ? nodes.children.map((node) => this.renderTreeItems(node)) : null
            }
        </TreeItem>);
    };

    render() {
        return (
            <TreeView defaultExpanded={['root']} defaultCollapseIcon={<FolderOpenIcon/>}
                      defaultExpandIcon={<FolderIcon/>} defaultEndIcon={<DescriptionIcon/>}
                      onNodeSelect={this.handleTreeItemClick} multiSelect>
                {this.renderTreeItems(this.state.policiesTreeData)}
            </TreeView>
        );
    }
}
