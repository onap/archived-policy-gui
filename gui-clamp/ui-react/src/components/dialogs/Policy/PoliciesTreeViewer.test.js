/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2022 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

import {shallow} from "enzyme";
import React from "react";
import PoliciesTreeViewer from "./PoliciesTreeViewer";
import fs from "fs";
import toJson from "enzyme-to-json";

describe('Verify PoliciesTreeViewer', () => {

    let toscaPoliciesData = fs.readFileSync('src/components/dialogs/Policy/toscaPoliciesData.test.json', {
        encoding: 'utf8',
        flag: 'r'
    });
    const toscaPoliciesDataArray = JSON.parse(toscaPoliciesData);

    const logSpy = jest.spyOn(console, 'log');


    it("renders correctly", () => {
        const component = shallow(<PoliciesTreeViewer policiesData={'[]'} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it("tests createPoliciesTree handler", () => {
        const component = shallow(<PoliciesTreeViewer policiesData={ toscaPoliciesDataArray } valueForTreeCreation="name" />);
        component.setState({ policiesTreeData: toscaPoliciesDataArray });

        const instance = component.instance();
        instance.createPoliciesTree(toscaPoliciesDataArray);

        expect(logSpy).toHaveBeenCalledWith('createPoliciesTree called');
        expect(component.state('policiesTreeData')).toEqual(toscaPoliciesDataArray);
    });

});