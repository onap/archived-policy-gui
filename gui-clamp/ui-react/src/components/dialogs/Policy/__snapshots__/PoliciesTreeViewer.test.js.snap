// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify PoliciesTreeViewer renders correctly 1`] = `
<WithStyles(ForwardRef(TreeView))
  defaultCollapseIcon={<Memo />}
  defaultEndIcon={<Memo />}
  defaultExpandIcon={<Memo />}
  defaultExpanded={
    Array [
      "root",
    ]
  }
  multiSelect={true}
  onNodeSelect={[Function]}
>
  <WithStyles(ForwardRef(TreeItem))
    key="1"
    label="ROOT(0)"
    nodeId={1}
    onNodeSelect={[Function]}
  />
</WithStyles(ForwardRef(TreeView))>
`;
